tensorflow/contrib/tensorrt/kernels/trt_calib_op.cc

```diff
@@ -0,0 +1,134 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/contrib/tensorrt/kernels/trt_calib_op.h"
+#include "tensorflow/contrib/tensorrt/resources/TRTInt8Calibrator.h"
+#include "tensorflow/contrib/tensorrt/resources/TRTResourceManager.h"
+#include "tensorflow/contrib/tensorrt/resources/TRTResources.h"
+#include "tensorflow/core/framework/node_def.pb.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/framework/types.h"
+
+#if GOOGLE_CUDA
+#if GOOGLE_TENSORRT
+#include "cuda_runtime_api.h"
+#include "tensorrt/include/NvInfer.h"
+
+namespace tensorflow {
+namespace trt {
+TRTCalibOp::TRTCalibOp(OpKernelConstruction* context) : OpKernel(context) {
+  OP_REQUIRES_OK(context, context->GetAttr("segment_nodes", &segment_nodes_));
+  OP_REQUIRES_OK(context, context->GetAttr("input_names", &input_names_));
+  OP_REQUIRES_OK(context, context->GetAttr("resource_name", &resource_name_));
+};
+
+//  case statement for type
+#define TYPECASE(dt, X, Y)                                             \
+  case dt: {                                                           \
+    Y = (void*)X->flat<tensorflow::EnumToDataType<dt>::Type>().data(); \
+    break;                                                             \
+  }
+
+// macro to get tensor data address pointed by tensor_ptr into
+// void* dest_ptr
+#define GET_TENSOR_ADDRESS(tensor_ptr, dest_ptr)               \
```

Comment:
Make this a function instead? Macro is hard to read/debug and sometimes errorprone.

Reply:
This is a big case statement. I saw same pattern in core files and that was the reason to use it.

Reply:
Are you talking about tensorflow/core/framework/tensor.cc? In that case we need the macro because we need to pass custom expressions to it (and it's not good and we can refactor it, too). In this case it looks like the arguments are pointers with fixed types (const Tensor*, void *), so please use a function instead.

Reply:
Removed the macro and put the code in place